Setting up Gatsby and Snipcart infrastructure

stefantul opened the pull request:
Open sesame

blackforestboi commented:
@stefantul thanks for your first PR :)

It will be good for the future to always have an opening comment on what is in the PR, and have a checklist for what is still to be done in order to close the PR. 
A good example of that: https://github.com/WorldBrain/Memex/pull/820

You can include all links for people to test as well. Thanks :)

ShishKabab commented:
Hi @stefantul , thanks for the work so far! Right now, I see you've directly embedded styles for prototyping purposes, but when seriously styling the website I'd like you to use [CSS modules](https://www.gatsbyjs.org/docs/css-modules/) so the code will stay easier to maintain. Aside from that, it'd be great if you could factor out interactions with Firebase to a dedicated class called `Services`, containing an `auth` property which does everything related to Auth. The idea is to keep our UI (input/rendering) and logic separated, which means nothing in React should know anything about Firebase directly. This makes refactoring easier and also automated testing when needed. You can see an example of this approach here:
https://github.com/WorldBrain/storex-frontend-boilerplate/blob/master/src/index.tsx#L16

stefantul commented:
Thanks, sound advice, will keep in mind for the next PR.

stefantul commented:
Checklist:

- add image to the top section of the homepage
- port to Gatsby & test Privacy Policy page
- port to Gatsby & test ToS page
- port to Gatsby & test Imprint page
- port to Gatsby & test FAQ's page
- port to Gatsby & test Price page, link with Snipcart
- check the feasibility of integrating the community section with Gatsby 
- remove unused js & css code pertaining to the ported WP plugins & theme
- fix console warnings & errors
- perform Lighthouse audit and perform fixes according to the results
- add Google analytics Gatsby plugin
- add SEO

blackforestboi commented:
> -  add Google analytics Gatsby plugin
> - add SEO

You can leave out those for now. 

> check the feasibility of integrating the community section with Gatsby

Can you elaborate what you mean here?

stefantul commented:
I'm referring to: https://worldbrain.io/community/myreferrals/ 
Gatsby produces static pages and static pages don't make use of external databases making the login/sign up more trickier to implement. 
https://www.keycdn.com/support/difference-between-static-and-dynamic

The Google Analytics and SEO are actually easy to configure, it won't take much to add them.

One more thing: which service does the "Subscribe to our monthly update email" feature uses?

blackforestboi commented:
> I'm referring to: https://worldbrain.io/community/myreferrals/

That does not need to be implemented. Just the upgrade history. 

> The Google Analytics and SEO are actually easy to configure, it won't take much to add them.

Due to privacy requirements we don't use Google Analytics. SEO we can implement later. Its definitely prio 2. 

> One more thing: which service does the "Subscribe to our monthly update email" feature uses?

That is mailchimp.

stefantul commented:
There are 2 things that need to be clarified: The community section for people who upgraded and the hosting. Are there people who created an account and upgraded? If so, that would mean moving/transferring somehow their existing accounts to Snipcart. 

Hosting: I recommend using Netlify instead of AWS Amplify, it's easy to configure, you just have to create a Netlify account and then import from git the repository as a source and Netlify will automatically build a new version of the website each time a new modification to the repository is detected. Netlify offers free hosting and CDN and integrates great together with Gatsby. Using AWS Amplify makes more sense when using and integrating together various AWS services.

blackforestboi commented:
> Are there people who created an account and upgraded? If so, that would mean moving/transferring somehow their existing accounts to Snipcart

Yes we have to transfer them from Wordpress/Woocommerce. We are likely letting them log in again because they need a new user ID etc to be saved. 

> Hosting: I recommend using Netlify instead of AWS Amplify,

What do you say @ShishKabab ?

stefantul commented:
A list of the things implemented in the new commit:

- added image to the top section of the homepage
- ported to React & Gatsby:  Privacy Policy/ToS/Imprint/Manifesto/Pricing/FAQ's pages
- converted old cart to Snipcart & configured the Pricing page
- new User page (replaces the Login page) where the user can access his Snipcart profile
- removed JS&CSS unused code coming from WP as much as possible
- integrated Mailchimp with Gatsby

Things to do:
- find a solution to transfer old user accounts from Woocommerce to Snipcart.
- perform Lighthouse audit and perform fixes according to the results (this should be done after the website is hosted because Lighthouse analyses parameters such as the page loading speed)
- add environment variables (such as Snipcart API key) to the hosting solution (Netlify has a nice way of doing it - https://www.netlify.com/docs/build-settings/)

blackforestboi commented on config.json:
```diff
@@ -0,0 +1,9 @@
+
+{
+	"apiKey": "AIzaSyDjS6tGpqjAaC0aXQAGZMkjssABbZzTgSo",
```
Ouch! can you put this in a gitignore file please? This should not be public!

ShishKabab replied:
Yes, this means the API key is compromised and should be regenerated. Once on the internet, always on the internet.

stefantul replied:
This config is already public and is part of the Gatsby & Firebase starter that I've been testing out: https://github.com/muhajirdev/gatsby-starter-firebase/blob/master/src/firebase.js 

But yes, API keys shouldn't be public.

ShishKabab commented:
> Hosting: I recommend using Netlify instead of AWS Amplify,

Personally, I'm torn. Netlify looks really developer-friendly with a lot of great features. But, also Amplify supports per-branch previews and hooks nicely into a repository. Reasons we chose for Amplify was pricing, to not use too many services at the same time, and to easily integrate with AWS services once we move higher maintanence, lower pricing setup than Firebase.

But, if my understanding of Amplify Git integration is wrong, or you have compelling reasons for Netlify over Amplify, I'd be happy to reconsider (since almost anything is more ethical than Amazon.)

stefantul commented:
Well, if using AWS for scaling is the end goal, why not start using them, eg. Cognito instead of Firebase.  AWS has a generous free tier plan which is ok solution for a startup.

ShishKabab commented:
We've thought about that, but the added complexity is not worth it for now for our other system components.

I see however you deployed something on Netlify without a clear agreement on moving there. Next time, please ping someone on Slack if you need quick feedback on an important decision.

blackforestboi commented:
I think we can close this one @ShishKabab. 

Superseeded by #2
